procrastinate/app.py

```diff
@@ -128,6 +130,9 @@ def _register(self, task: "tasks.Task") -> None:
             )
             self.queues.add(task.queue)
 
+    def _register_builtin_tasks(self) -> None:
+        self._register(self.task(builtin_tasks.remove_old_jobs, queue="builtin"))
```

Comment:
I'd rather call a function defines in the builtin_tasks module here (with the app parameter).

Also, I'd rather call `app.task` than register.

I'm guessing the problem is that you don't yet know the app instance when you define the builtin tasks, but we could have:

```python
def my_builtin():
	...

BUILTIN_TASKS_FUNCS = [
	my_builtin,
]
def register_builtins(app: ...) -> None:
	for func in BUILTIN_TASKS_FUNCS:
		app.task(func)
```

Finally, a test would check that all the functions defined in this module (except a few ones) are in BUILTIN_TASKS_FUNCS, so that the risk of forgetting one would be low ?

Reply:
Even better:
```python
app.task(functools.partial(func, app=app))
```

Maybe with

```python
@pass_app
def my_builtin():
	...
```

This way, your job can access the app, and thus the job store, and thus make a query.

Reply:
The issue is indeed that you can't access the app instance when defining the builtin task :o

> Also, I'd rather call app.task than register.

Isn't it necessary to `register` the task to call it by CLI later on?

I tried `functools.partial` and it fricks up the name resolution later on. The root issue being that a `partial` doesn't have a path. Maybe there is a better solution than what I did, or we could try to power through.

Reply:
>> Also, I'd rather call app.task than register.
> Isn't it necessary to register the task to call it by CLI later on? (This is not my main issue, I'm trying to understand this particular remark :>)

Oh, I had not noticed that you're actually calling `self.task` (which is `app.task`). This method instantiates a task *and* registers it, so no need to call `_register` yourself, just call self.task as you did.

> I tried functools.partial and it fricks up the name resolution later on. The root issue being that a partial doesn't have a path.

This is not a problem if you define the task name explicitely (`name=...`). And you're free to have this name be the builtin path you'd expect. As long as a task is registered, its name doesn't have to be it's path. The dynamic import is only done for unregistered tasks.

Reply:
Ok for `self.task` :+1: 

You can give a `name` if you want, but the code will still check for the `full_path`: https://github.com/peopledoc/procrastinate/blob/master/procrastinate/tasks.py#L39 And it crashes since a partial does not have a path :o
(It crashes further down if you comment this block)

Not saying it's a deal breaker or anything, but it's less simple than you make it out to be :>

Reply:
yeah, sorry about that.

I think the crashing is the problem. One should be able to register a task that has no path. I'd say we can remove the warning. And fix whatever crashed further down ?

Reply:
Seems like a good call yes :) I'll look into it on Friday